torch/csrc/dynamo/eval_frame.c

```diff
@@ -0,0 +1,578 @@
+#else
+  if (tstate->interp->eval_frame != &_PyEval_EvalFrameDefault) {
+    // First call
+    tstate->interp->eval_frame = &_PyEval_EvalFrameDefault;
+  }
+#endif
+}
+
+static inline PyObject* call_callback(
+    PyObject* callable,
+    PyObject* frame,
+    long cache_len) {
+  PyObject* args = Py_BuildValue("(Ol)", frame, cache_len);
+  NULL_CHECK(args);
+  PyObject* result = PyObject_CallObject(callable, args);
+  Py_DECREF(args);
+  return result;
+}
+
+typedef struct cache_entry {
+  // check the guards: lambda: <locals of user function>: bool
+  PyObject* check_fn;
+  // modified user bytecode (protected by check_fn's guards)
+  PyCodeObject* code;
+  // on a cache miss, linked list of next thing to try
+  struct cache_entry* next;
+} CacheEntry;
+
+static CacheEntry* create_cache_entry(
+    CacheEntry* next,
+    PyObject* guarded_code) {
+  CacheEntry* e = (CacheEntry*)malloc(sizeof(CacheEntry));
+  DEBUG_NULL_CHECK(e);
+  e->check_fn = PyObject_GetAttrString(guarded_code, "check_fn");
+  NULL_CHECK(e->check_fn);
+  e->code = (PyCodeObject*)PyObject_GetAttrString(guarded_code, "code");
+  NULL_CHECK(e->code);
+  e->next = next;
+  return e;
+}
+
+static void destroy_cache_entry(CacheEntry* e) {
+  if (e == NULL || e == SKIP_CODE) {
+    return;
+  }
+  Py_XDECREF(e->check_fn);
+  Py_XDECREF(e->code);
+  destroy_cache_entry(e->next);
+  free(e);
+}
+
+#ifdef TORCHDYNAMO_DEBUG
+inline static const char* name(PyFrameObject* frame) {
+  DEBUG_CHECK(PyUnicode_Check(frame->f_code->co_name));
+  return PyUnicode_AsUTF8(frame->f_code->co_name);
+}
+#endif
+
+static void call_guard_fail_hook(
+    PyObject* hook,
+    CacheEntry* e,
+    PyObject* f_locals) {
+  // call debugging logic when a guard fails
+  PyObject* args = PyTuple_Pack(
+      4,
+      e->check_fn,
+      e->code,
+      f_locals,
+      (e->next == NULL ? Py_True : Py_False));
+  NULL_CHECK(args);
+  PyObject* result = PyObject_CallObject(hook, args);
+  NULL_CHECK(result);
+  Py_DECREF(result);
+  Py_DECREF(args);
+}
+
+static PyCodeObject* lookup(CacheEntry* e, PyObject* f_locals) {
+  if (e == NULL) {
+    return NULL;
+  }
+  PyObject* dotzero = PyDict_GetItem(f_locals, dotzerokey);
+  PyObject* valid = NULL;
+  if (unlikely(dotzero != NULL)) {
+    // .0 is a special variable name used for implicit args
+    PyObject* args = PyTuple_Pack(1, dotzero);
+    NULL_CHECK(args);
+    valid = PyObject_Call(e->check_fn, args, f_locals);
+    Py_DECREF(args);
+  } else {
+    valid = PyObject_Call(e->check_fn, noargs, f_locals);
+  }
+  if (unlikely(valid == NULL)) {
+    PyErr_Print();
+    if (guard_error_hook != NULL) {
+      call_guard_fail_hook(guard_error_hook, e, f_locals);
+    }
+    NULL_CHECK(valid);
+  }
+  Py_DECREF(valid);
+  if (valid == Py_True) {
+    return e->code;
+  }
+  if (unlikely(guard_fail_hook != NULL)) {
+    call_guard_fail_hook(guard_fail_hook, e, f_locals);
+  }
+  return lookup(e->next, f_locals);
+}
+
+static long cache_size(CacheEntry* e) {
```

Comment:
This seems to be in the very hot path and the value can easily be cached?

Reply:
Seems reasonable to cache it.  This shouldn't be on the hot path though, it is only called on cache misses.